### ISSUE TITLE
Bulk operations do not scale

### ISSUE THREAD

**Issue by lampwins** [nautobot/nautobot#2459]
### Environment
* Nautobot version (Docker tag too if applicable): 1.4.3
* Python version: 3.9
* Database platform, version: All
* Middleware(s): N/A

### Steps to Reproduce
1. Have a large data set with many thousands of devices and/or other objects (>50,000)
2. Attempt to perform a bulk update or bulk delete on a large subset of the inventory (~2000) in the UI. For example, updating a custom field value, changing tags, deleting devices, etc.
3. Wait

### Expected Behavior
The operation should succeed and preferably in a reasonable amount of time and with continual feedback to the user.

### Observed Behavior
The operation is contained within a single HTTP request but this request takes a very long time to process, so the user gets no feedback that Nautobot is actually doing anything. Depending on the size of the operation, the request will eventually time out and the user is presented with an ugly 504. I call out the UI specifically here, but the same fundamental issue exists with the REST API.

**Comment by lampwins** [comment]
Bug feature :)

**Comment by travisjohnsonga** [comment]
I have this issue in 2.0.1

Trying to delete ~1200 devices results in 504 gateway time out

**Comment by lampwins** [comment]
We will primarily consider bulk update and delete cases. Need to handle the case where the user selects all records, and when only a subset (a large subset) is selected.

Needs to target 2.x and 1.6.

**Comment by glennmatthews** [comment]
Ideas to pursue:

- Profiling suggests that change logging overhead appears to be a significant contributor, perhaps with DRF as a key factor of that in turn
  - PRs like #5293 might be a direction to pursue?
  - Backport #5329 to LTM?
- With the need to target 1.6 as well as 2.x, may need a split set of approaches? E.g. we could move these to system jobs in 2.x but not in 1.6, etc.
- See #5321 - for bulk delete in particular, when "all" is selected, we should be able to pass around the queryset and/or filter params instead of constructing a `pk_list` to render the confirmation page. May be a factor with bulk edit as well.
- Possibly look at auto-optimizing querysets on edit/delete to improve change logging performance, e.g. take some of the code from #5215 and generalize it as a queryset helper?

**Comment by gsnider2195** [comment]
Potential crazy idea

```py
with transaction.atomic():
    # disable change logging
    # do the change logging manually so we can leverage the select_related fields
    # try to delete the queryset
    # enable change logging
```

**Comment by gsnider2195** [comment]
Testing with silk profiling in 1.6:

With change logging enabled, deleting 115 prefixes, the delete performed ~1200 SQL queries and took around 13 seconds

With change logging disabled globally via middleware (this also disables webhook/jobhook processing), it was reduced to 12 queries and <1 second

**Comment by glennmatthews** [comment]
#5329 appears unnecessary to backport to ltm-1.6 because there `enqueue_job_hooks` is [using](https://github.com/nautobot/nautobot/blob/ltm-1.6/nautobot/extras/jobs.py#L1349-L1352) `ChangeLoggedModelsQuery().list_subclasses()`, which doesn't involve a database hit in the first place. I did an experimental cherry-pick of this PR into ltm-1.6 anyway, and it appeared to have negligible impact on the performance of bulk-delete or bulk-update operations.

**Comment by lampwins** [comment]
Sounds like we start with manually processing changelog out of band of bulk operations. Potentially overriding bulk_update and bulk_delete on RestrictedQueryset.

**Comment by lampwins** [comment]
Related to #5321, this likely all hinges on the logic in the bulk views.

**Comment by gsnider2195** [comment]
I did some performance testing as part of #5529 and it looks like we have some opportunities to explore why our number of sql queries is so high and why our serializers are so slow. I tried an approach where we implemented a generic batching of bulk create/update for objectchanges [here](https://github.com/nautobot/nautobot/commit/c229dcbf961d800da72dc069c8d289cf69b47802) and that's shown in the testing below as the branch `u/gas-2459-bulk-edit-delete`. 

TLDR: The only consistent improvement I saw was from removing the serializer calls.

## Testing results

### Test code

```py
import ipaddress
import time
import tracemalloc

from django.contrib.contenttypes.models import ContentType
from django.db import connection
from django.test import override_settings
from django.test.utils import CaptureQueriesContext
from django.urls import reverse

from nautobot.core.testing import post_data, TransactionTestCase
from nautobot.extras.models import ObjectChange, Status
from nautobot.ipam.choices import PrefixTypeChoices
from nautobot.ipam.models import Namespace, Prefix
from nautobot.users.models import ObjectPermission


class PrefixBulkEditTestCase(TransactionTestCase):
    model = Prefix

    @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"])
    def test_bulk_edit_performance(self):
        test_namespace = Namespace.objects.create(name="test_performance_namespace")
        pfx_status = Status.objects.get_for_model(Prefix).first()
        pk_list = []
        for i in range(100):
            pk_list.append(
                Prefix.objects.create(
                    network=ipaddress.ip_address(i),
                    prefix_length=32,
                    type=PrefixTypeChoices.TYPE_NETWORK,
                    status=pfx_status,
                    namespace=test_namespace,
                ).pk
            )
        data = {
            "pk": pk_list,
            "_apply": True,  # Form button
        }

        # Append the form data to the request
        data.update(post_data({"description": "test performance"}))

        # Assign model-level permission
        obj_perm = ObjectPermission(name="Test permission", actions=["change"])
        obj_perm.save()
        obj_perm.users.add(self.user)
        obj_perm.object_types.add(ContentType.objects.get_for_model(self.model))

        object_changes = ObjectChange.objects.count()

        url = reverse("ipam:prefix_bulk_edit")

        def runtests():
            tests = []
            for i in range(10):
                start_time = time.time()
                self.assertHttpStatus(self.client.post(url + f"?namespace={test_namespace.name}", data), 302)
                end_time = time.time()
                tests.append(end_time - start_time)

            return tests

        # starting memory monitoring - only use this when testing memory consumption as it tanks performance
        tracemalloc.start()

        with CaptureQueriesContext(connection) as queries:
            tests = runtests()

        # retrieve memory used
        memory_size, memory_peak = tracemalloc.get_traced_memory()

        # stopping memory monitoring
        tracemalloc.stop()

        object_changes_created = ObjectChange.objects.count() - object_changes
        self.fail(
            "Bulk edit performance test took on average {:.2f} seconds, {} object changes created, {} sql queries performed, using {} bytes of memory".format(
                sum(tests) / len(tests),
                object_changes_created,
                len(queries),
                memory_peak,
            )
        )
```

### bulk editing 100 prefixes

- `develop`
    - 2019 sql queries performed
- `develop` with `object_data` and `object_data_v2` set to `{}` in `ChangeLoggedModel.to_objectchange`
    - 1417 sql queries performed
- `u/timizuoebideri-2459-bulk-operations-do-not-scale`
    - 1820 sql queries performed
- `u/gas-2459-bulk-edit-delete`
    - 1920 sql queries performed

### bulk editing 100 prefixes 10 times

- `develop`
    - test took on average 3.16 seconds, 1000 object changes created, using 9970966 bytes of memory
- `develop` with `object_data` and `object_data_v2` set to `{}` in `ChangeLoggedModel.to_objectchange`
    - test took on average 2.14 seconds, 1000 object changes created, using 2282065 bytes of memory
- `u/timizuoebideri-2459-bulk-operations-do-not-scale`
    - test took on average 2.92 seconds, 1000 object changes created, using 38024120 bytes of memory
- `u/gas-2459-bulk-edit-delete`
    - test took on average 3.04 seconds, 1000 object changes created, using 34835637 bytes of memory

### bulk editing 1000 prefixes 10 times

- `develop`
    - test took on average 43.59 seconds, 10000 object changes created, using 22851096 bytes of memory
- `develop` with `object_data` and `object_data_v2` set to `{}` in `ChangeLoggedModel.to_objectchange`
    - test took on average 21.78 seconds, 10000 object changes created, using 8717534 bytes of memory
- `u/timizuoebideri-2459-bulk-operations-do-not-scale`
    - test took on average 37.31 seconds, 10000 object changes created, using 155468072 bytes of memory
- `u/gas-2459-bulk-edit-delete`
    - test took on average 31.51 seconds, 10000 object changes created, using 155367224 bytes of memory

**Comment by gsnider2195** [comment]
I wasn't happy with my test results above since they were so far off from @timizuoebideri1's testing results in #5529 so I ran these tests again with everything else on my machine closed, all profiling turned off, including pycoverage, but this time on 10k prefixes and the results were pretty similar:

### bulk editing 10000 prefixes 10 times -- COVERAGE DISABLED

![image](https://github.com/nautobot/nautobot/assets/75227981/bbce7be1-e6d3-4de2-b7fd-5ced86e6f89f)

- `develop`
    - `257.14593529701233`
    - `263.88606333732605`
    - `301.02319955825806`
    - `263.4733431339264`
    - `299.40543842315674`
    - `262.21188020706177`
    - `264.42850065231323`
    - `268.4449782371521`
    - `287.02962350845337`
    - `309.2388799190521`
- `develop` with `object_data` and `object_data_v2` set to `{}` in `ChangeLoggedModel.to_objectchange`
    - `177.58864068984985`
    - `187.78214597702026`
    - `181.50266885757446`
    - `160.60235905647278`
    - `151.17761325836182`
    - `149.07634472846985`
    - `152.33581566810608`
    - `150.3669776916504`
    - `162.4092400074005`
    - `155.99419951438904`
- `u/timizuoebideri-2459-bulk-operations-do-not-scale`
    - `267.15529704093933`
    - `270.5580883026123`
    - `259.71550393104553`
    - `278.2030391693115`
    - `298.38112020492554`
    - `249.15813302993774`
    - `267.0685520172119`
    - `267.62854623794556`
    - `262.05445075035095`
    - `257.0511803627014`
- `u/gas-2459-bulk-edit-delete`
    - `245.6239058971405`
    - `251.31356573104858`
    - `253.4888219833374`
    - `249.80808353424072`
    - `249.56358242034912`
    - `272.4603054523468`
    - `247.49774837493896`
    - `257.469881772995`
    - `253.8151092529297`
    - `249.16804456710815`

**Comment by gsnider2195** [comment]
Notes from design review discussion:

We still have some room for improvement on our select_related and prefetch_related on the bulk views. We already have some automatic logic to determine the correct select_related and prefetch_related fields on [`BaseTable.__init__`](https://github.com/nautobot/nautobot/blob/a1b9dd49f9489c94f7af130862f9d0a9a62cc909/nautobot/core/tables.py#L37). As a first pass we should extend this logic to any other view that benefits from select_related and prefetch_related, prioritizing bulk edit and bulk delete and the NautobotUIViewset equivalents.

Another idea to explore is using the `object_data` field (fast django serialization of just the instance data - ie no table joins) to determine the actual fields that were changed and then only serializing those fields in the `object_data_v2` data. We'll want some input from @lampwins on this idea.

**Comment by gsnider2195** [comment]
> determine the actual fields that were changed and then only serializing those fields in the object_data_v2 data

After discussing with @lampwins this doesn't look like it's an option because of the expectation that ObjectChange instances will contact the full representation of the model.

**Comment by lampwins** [comment]
Closing this in favor of #6353 and #6354. There is good comentary in here for future actions, but moving these UX workflows to background processing should be a first step no matter what as it will still be needed at large enough scales.
